### PR TITLE
Towards SoftAdapt loss balancing for tf.compat.v1

### PR DESCRIPTION
**Pull request by pescap**
Work in progress!

### CODE REVIEW

**Comment by lululxvi** [comment]
Implementing in TensorFlow is tricky, as it is static graph. It should be much easier to implement in pytorch, where you can directly change the `loss_weights` value.

File: deepxde/callbacks.py [hunk 1]
```diff
@@ -571,3 +571,34 @@ def on_epoch_end(self):
...
+        super().__init__()
+
+        self.beta = beta
+        self.epsilon = epsilon
+
+    def on_train_begin(self):
+        loss_weights = tf.constant(self.model.loss_weights)
+        loss_weights = dde.Variable(loss_weights, trainable=False, dtype=loss_weights.dtype)
+        loss_weights *= 0
+
```
[the remark this answers is not in the document]
Comment (lululxvi) [reply deepxde/callbacks.py]: How do you plan to update `loss_weights`?

**Comment by haison19952013** [comment]
> Implementing in TensorFlow is tricky, as it is static graph. It should be much easier to implement in pytorch, where you can directly change the `loss_weights` value.

- It can be done if the loss_weights is the argument of the [train_step()](https://github.com/lululxvi/deepxde/blob/85920299331bd7c0bad01f3d2abba442a77c89c6/deepxde/model.py#L240)
- The problem is if we iteratively change the loss weights, `tensorflow` will have to [make the graph all over again](https://www.tensorflow.org/guide/intro_to_graphs#performance_and_trade-offs) and the training might be slow

**Comment by pescap** [comment]
Hi, if we define `loss_weights` as `Variable`, no need to compile several times, right?

Next, we have to define appopriately the [total_loss](https://github.com/lululxvi/deepxde/blob/85920299331bd7c0bad01f3d2abba442a77c89c6/deepxde/model.py#L244).

**Comment by haison19952013** [comment]
> Hi, if we define `loss_weights` as `Variable`, no need to compile several times, right?
> 
> Next, we have to define appopriately the [total_loss](https://github.com/lululxvi/deepxde/blob/85920299331bd7c0bad01f3d2abba442a77c89c6/deepxde/model.py#L244).

- My last response might raise some confusion so I have corrected it.
- You are right, we can update it like that without recompiling
- But, it's better to verify this again
- For more details, we can put loss_weights as the arguments of [outputs_losses()](https://github.com/lululxvi/deepxde/blob/85920299331bd7c0bad01f3d2abba442a77c89c6/deepxde/model.py#L205)

**Comment by PillowTomZA** [comment]
Hi @pescap @lululxvi @haison19952013 

I have recently been working on the adaptive weights, and here's what worked for me (I think).
Take this code with a big grain of salt, but do let me know what you think. 
I've adapted this from the code I see here, and have used this for my code (which has 3 governing equation losses, 8 boundary losses, and 3 data losses)

```python
from deepxde.backend import tf, Variable

class SoftAdapt(dde.callbacks.Callback):
    """Use adaptive loss balancing.
    Args:
        beta: If beta > 0, then softAdapt will pay more attention the worst performing
            loss component. If beta < 0, then SoftAdapt will assign higher weights
            to the better performing components. Beta==0 is the trivial case and
            all loss components will have coefficient 1.
        epsilon: parameter to prevent overflows.
    """
    def __init__(self, beta=0.1, epsilon=1e-8,losshistory=None,lr=0.01):
        super().__init__()
        
        self.beta = beta
        self.epsilon = epsilon
        self.losshistory=losshistory
#         self.epoch = 0
        self.lr = lr
        
    def on_epoch_end(self):

        loss_weights = self.losshistory.loss_weights
        weight_bc = loss_weights[3]
        weight_data = loss_weights[-1]
        current_loss = self.losshistory.loss_train[-1]

        pde_avg = current_loss[0:3].mean()

        bc_avg = current_loss[3:11].mean()

        data_avg = current_loss[11::].mean()

        weight_bc = (1-self.lr) * weight_bc + self.lr * current_loss[0:3].max() / bc_avg
        weight_data = (1 - self.lr) * weight_data + self.lr * current_loss[0:3].max() / data_avg

        loss_weights = [1] * 3 + [weight_bc] * 8 + [weight_data] * 3
        self.losshistory.set_loss_weights(loss_weights)

        print(self.losshistory.loss_weights, "loss_weights")
```

**Comment by lululxvi** [comment]
What is the backend you have tested?

**Comment by PillowTomZA** [comment]
I've used tensorflow.v1 compact for this code

**Comment by lululxvi** [comment]
I don't think the code would change the weights during training. For TF v1, if you don't recompile, it always uses the original computational graph.

**Comment by echen5503** [comment]
@ThomasHuangZA @pescap is this still being worked on?

**Comment by ThomasHuangZA** [comment]
Hi @echen5503 , thanks for the message. I am not working on this at the moment.
